### PR TITLE
rustup

### PR DESCRIPTION
**Pull request by globin**
This temporarily allows warnings for test to keep calls to `CString::from_{vec,slice}` for now as this creates an enormous fallout which is probably mitigated partly by a switch to the new io, fs, path libraries as far as I can see.

### CODE REVIEW

**Comment by globin** [comment]
Depends on carllerche/hamcrest-rust#15

**Comment by jakub-** [comment]
This should be good to go, I think?

File: src/cargo/lib.rs [hunk 1]
```diff
@@ -1,6 +1,7 @@
 #![deny(unused)]
-#![feature(collections, hash, io, libc, os, path, std_misc, unicode, env, core)]
-#![cfg_attr(test, deny(warnings))]
+#![feature(collections, hash, libc, os, std_misc, unicode, env, core)]
+#![feature(old_io, old_path, str_words)]
+//#![cfg_attr(test, deny(warnings))]
```
Comment (alexcrichton) [inline src/cargo/lib.rs]: Can this be re-added?

File: src/cargo/ops/cargo_rustc/job.rs [hunk 2]
```diff
@@ -21,7 +21,7 @@ impl<A, R, F: FnOnce(A) -> R> FnBox<A, R> for F {
 }
 
 impl Work {
-    pub fn new<F>(f: F) -> Work
+    pub fn new<F: 'static>(f: F) -> Work
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_rustc/job.rs]: Can you add this bound to the `where` clause below?

File: src/cargo/util/errors.rs [hunk 3]
```diff
@@ -58,8 +59,8 @@ impl<'a, T, F> ChainError<T> for F where F: FnOnce() -> CargoResult<T> {
     }
 }
 
-impl<T, E: CargoError> ChainError<T> for Result<T, E> {
-    fn chain_error<E2, C>(self, callback: C) -> CargoResult<T>
+impl<T, E: CargoError + 'static> ChainError<T> for Result<T, E> {
```
Comment (alexcrichton) [inline src/cargo/util/errors.rs]: Could you add this `'static` bound to `CargoError` itself?

File: tests/support/mod.rs [hunk 4]
```diff
@@ -362,6 +363,7 @@ fn lines_match(expected: &str, mut actual: &str) -> bool {
 struct ZipAll<T, I1, I2> {
     first: I1,
     second: I2,
+    marker: PhantomData<T>,
```
Comment (alexcrichton) [inline tests/support/mod.rs]: This can actually be removed entirely by placing `Iterator` bounds on `I1` and `I2` and then the type can go away.

Comment (alexcrichton) [inline tests/support/mod.rs]: (the type parameter `T` that is)

**Comment by alexcrichton** [comment]
Looks like the `.travis.install.deps.sh` script needs to be updated as well as I think the structure of the tarballs has changed.

File: src/cargo/lib.rs (hunk 1 above)
Comment (globin) [inline src/cargo/lib.rs]: There are deprecation warnings with `CString::from_{vec,slice}` that are not easily fixable and would need quite a lot of work currently and I reckon that would be easier with the new io/path/os/fs in plac.

**Comment by huonw** [comment]
Incidentally, it would be great if this patch could compile with alpha.2 (as written `cargo build` succeeds with it*). It would help with @anguslees's rust/cargo packaging experiments.

(*but `cargo test` fails to compile hamcrest.)

File: src/cargo/lib.rs (hunk 1 above)
Comment (globin) [inline src/cargo/lib.rs]: And just tried with alpha.2 which also fails because the unstable features create a warning making the test build fail.

**Comment by globin** [comment]
Updated, have I introduced the failing tests?

**Comment by alexcrichton** [comment]
There are still some travis failures (cross builds and such). I think the installation format has changed and the hokey merge-the-repos behavior we had before isn't quite so valid any more. I also don't mind taking this over to help push it through.

**Comment by globin** [comment]
Thanks for taking over :+1:

**Pull request merged**